From 2a22a1e6be04e9d38948180142f3d90e81b691db Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Fri, 4 Oct 2024 14:33:21 +0200
Subject: [PATCH 690/954] feat(stm32mp2): force autonomous mode for LPSRAM1/2/3
 when CPU3 is used

Force the autonomous mode for LPSRAM1/2/3 with AMEN bit in RCC_LPSRAMxCFGR
when CPU3=M0 is running and when it is needed for low power mode,
in Standby or in LPLV-Stop because its firmware is running in LPSRAMx.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 11d21feae7cb62beee17940bd097d148f5fe4fb7 feat(stm32mp2): force autonomous mode for LPSRAM1/2/3 when CPU3 is used]
Change-Id: I4fb272e41b09dc77ace0c2bcadefba8a0efb134e
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/405969
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/stm32mp2_pm.c | 65 ++++++++++++++++++++++++++++++++++
 1 file changed, 65 insertions(+)

diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index c36d1e891e..cf6b848d42 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -91,6 +91,11 @@ static u_register_t saved_scr_el3;
 static uint32_t lpstop1_pwrlpdly;
 static uint32_t stop2_pwrlpdly;
 
+#if !STM32MP21
+/* bitfield to indicate the modified AMEN bit for LP-SRAM1/2/3 */
+static uint32_t saved_lpsram_amen;
+#endif
+
 /* Support PSCI v1.0 Extended State-ID with the recommended encoding */
 #define LVL_CORE		U(0)
 #define LVL_D1			U(1)
@@ -239,6 +244,61 @@ static void stm32mp_ca35_lpi_restore(void)
 	}
 }
 
+/* LPSRAM1/2/3 autonomous support (AMEN) when D3 is used in low power */
+static void lpsram_autonomous_mode_set(uintptr_t rcc_base)
+{
+#if !STM32MP21 /* STM32MP21 Series don't have D3 domain */
+	const uint32_t mask = RCC_C3CFGR_C3EN | RCC_C3CFGR_C3AMEN;
+
+	/* configure LPSRAM only when C3 is used in autonoumous mode */
+	if ((mmio_read_32(rcc_base + RCC_C3CFGR) & mask) != mask) {
+		return;
+	}
+
+	saved_lpsram_amen = 0U;
+
+	/* force autonomous mode (ANEN) when clock is enabled (EN) and ANEN is not yet set */
+	if ((mmio_read_32(rcc_base + RCC_LPSRAM1CFGR) &
+	     (RCC_LPSRAM1CFGR_LPSRAM1EN | RCC_LPSRAM1CFGR_LPSRAM1AMEN)) ==
+	      RCC_LPSRAM1CFGR_LPSRAM1EN) {
+		mmio_setbits_32(rcc_base + RCC_LPSRAM1CFGR,
+				RCC_LPSRAM1CFGR_LPSRAM1AMEN);
+		saved_lpsram_amen |=  BIT(0);
+	}
+	if ((mmio_read_32(rcc_base + RCC_LPSRAM2CFGR) &
+	     (RCC_LPSRAM2CFGR_LPSRAM2EN | RCC_LPSRAM2CFGR_LPSRAM2AMEN)) ==
+	      RCC_LPSRAM2CFGR_LPSRAM2EN) {
+		mmio_setbits_32(rcc_base + RCC_LPSRAM2CFGR,
+				RCC_LPSRAM2CFGR_LPSRAM2AMEN);
+		saved_lpsram_amen |=  BIT(1);
+	}
+	if ((mmio_read_32(rcc_base + RCC_LPSRAM3CFGR) &
+	     (RCC_LPSRAM3CFGR_LPSRAM3EN | RCC_LPSRAM3CFGR_LPSRAM3AMEN)) ==
+	      RCC_LPSRAM3CFGR_LPSRAM3EN) {
+		mmio_setbits_32(rcc_base + RCC_LPSRAM3CFGR,
+				RCC_LPSRAM3CFGR_LPSRAM3AMEN);
+		saved_lpsram_amen |=  BIT(2);
+	}
+#endif /* !STM32MP21 */
+}
+
+static void lpsram_autonomous_mode_restore(uintptr_t rcc_base)
+{
+#if !STM32MP21
+	if (saved_lpsram_amen & BIT(0))
+		mmio_clrbits_32(rcc_base + RCC_LPSRAM1CFGR,
+				RCC_LPSRAM1CFGR_LPSRAM1AMEN);
+	if (saved_lpsram_amen & BIT(1))
+		mmio_clrbits_32(rcc_base + RCC_LPSRAM2CFGR,
+				RCC_LPSRAM2CFGR_LPSRAM2AMEN);
+	if (saved_lpsram_amen & BIT(2))
+		mmio_clrbits_32(rcc_base + RCC_LPSRAM3CFGR,
+				RCC_LPSRAM3CFGR_LPSRAM3AMEN);
+
+	saved_lpsram_amen = 0U;
+#endif /* !STM32MP21 */
+}
+
 /*******************************************************************************
  * STM32MP2 handler called when a CPU is about to enter standby.
  ******************************************************************************/
@@ -518,6 +578,7 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 
 	case PWRSTATE_LPLV_STOP1:
 		print_mode_verbose("LPLV_Stop1");
+		lpsram_autonomous_mode_set(rcc_base);
 		mmio_write_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_LPDS_D1 | PWR_CPU1CR_LVDS_D1);
 		if (!cpu2_running) {
 			mmio_write_32(pwr_base + PWR_CPU2CR,
@@ -552,6 +613,7 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 
 	case PWRSTATE_LPLV_STOP2:
 		print_mode_info("LPLV_Stop2");
+		lpsram_autonomous_mode_set(rcc_base);
 		mmio_write_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_PDDS_D1);
 		if (!cpu2_running) {
 			mmio_write_32(pwr_base + PWR_CPU2CR,
@@ -565,6 +627,7 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 
 	case PWRSTATE_STANDBY:
 		print_mode_info("Standby1");
+		lpsram_autonomous_mode_set(rcc_base);
 		mmio_write_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_PDDS_D1 | PWR_CPU1CR_PDDS_D2);
 		if (!cpu2_running) {
 			mmio_write_32(pwr_base + PWR_CPU2CR, PWR_CPU2CR_PDDS_D2);
@@ -718,6 +781,8 @@ static void stm32_pwr_domain_suspend_finish(const psci_power_state_t
 	/* Restore SCR */
 	write_scr_el3(saved_scr_el3);
 
+	lpsram_autonomous_mode_restore(rcc_base);
+
 	/* Core synchronization to protect DDR access */
 	stm32mp_state_set(STM32MP_PRIMARY_CPU, STATE_DDR, true);
 	/* Unblock the other core in CPU standby loop */
-- 
2.53.0

